interfaces/prompting/patterns.go

```diff
@@ -0,0 +1,116 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"fmt"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+// Limit the number of groups which are allowed to occur in a pattern.
+// This number includes all groups, including those nested or in series.
+const maxGroupsInPattern = 10
+
+// ValidatePathPattern returns nil if the pattern is valid, otherwise an error.
+func ValidatePathPattern(pattern string) error {
+	if !doublestar.ValidatePattern(pattern) {
+		return fmt.Errorf("invalid path pattern: %q", pattern)
+	}
+	if pattern == "" || pattern[0] != '/' {
+		return fmt.Errorf("invalid path pattern: must start with '/': %q", pattern)
+	}
+	depth := 0
+	totalGroups := 0
+	reader := strings.NewReader(pattern)
+	for {
+		r, _, err := reader.ReadRune()
+		if err != nil {
+			// No more runes
+			break
+		}
+		switch r {
+		case '{':
+			depth += 1
+			totalGroups += 1
+			if totalGroups > maxGroupsInPattern {
+				return fmt.Errorf("invalid path pattern: exceeded maximum number of groups (%d): %q", maxGroupsInPattern, pattern)
+			}
+		case '}':
+			depth -= 1
+			if depth < 0 {
+				return fmt.Errorf("invalid path pattern: unmatched '}' character: %q", pattern)
+			}
+		case '\\':
+			// Skip next rune
+			_, _, err = reader.ReadRune()
```

Comment:
isn't the escape sequence valid only for some of the characters that follow? eg what would a pattern like `\c` mean?

Reply:
According to `doublestar.Match` (https://pkg.go.dev/github.com/bmatcuk/doublestar#Match), `\c` matches `c` for any character `c`. I'm not adjusting that behavior at all, at the moment.